File: c/src/fmhy.c
// SPDX-License-Identifier: MIT
// FMHY content catalogue — see aethernet/fmhy.h. Hand-rolled markdown parser
// (no regex dependency) mirroring the C# reference grammar.

#include "aethernet/fmhy.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// ─── small helpers ───────────────────────────────────────

static bool is_ws(char c) { return c == ' ' || c == '\t' || c == '\r' || c == '\n'; }

// Duplicate src[0..n) with leading/trailing ASCII whitespace trimmed.
static char *trim_ndup(const char *src, size_t n) {
    size_t start = 0, end = n;
    while (start < end && is_ws(src[start])) start++;
    while (end > start && is_ws(src[end - 1])) end--;
    size_t len = end - start;
    char *out = (char *)malloc(len + 1);
    if (!out) return NULL;
    memcpy(out, src + start, len);
    out[len] = '\0';
    return out;
}

static char *str_dup_fmhy(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s) + 1;
    char *out = (char *)malloc(n);
    if (out) memcpy(out, s, n);
    return out;
}

static bool starts_with(const char *s, const char *prefix) {
    return strncmp(s, prefix, strlen(prefix)) == 0;
}

static char to_lower_ascii(char c) { return (c >= 'A' && c <= 'Z') ? (char)(c - 'A' + 'a') : c; }

// Case-insensitive substring test (ASCII fold).
static bool contains_ci(const char *hay, const char *needle) {
    if (!hay || !needle || !*needle) return true;
    size_t nl = strlen(needle);
    for (const char *h = hay; *h; h++) {
        size_t i = 0;
        while (i < nl && h[i] && to_lower_ascii(h[i]) == to_lower_ascii(needle[i])) i++;
        if (i == nl) return true;
    }
    return false;
}

// ─── entry lifecycle ─────────────────────────────────────

static void free_entry_fields(aethernet_fmhy_entry_t *e) {
    free(e->name);
    free(e->url);
    free(e->description);
    free(e->category);
    for (int32_t i = 0; i < e->mirror_count; i++) free(e->mirrors[i]);
    free(e->mirrors);
    e->name = e->url = e->description = e->category = NULL;
    e->mirrors = NULL;
    e->mirror_count = 0;
}

void aethernet_fmhy_entries_free(aethernet_fmhy_entry_t *entries, int32_t count) {
    if (!entries) return;
    for (int32_t i = 0; i < count; i++) free_entry_fields(&entries[i]);
    free(entries);
}

static aethernet_fmhy_entry_t copy_entry(const aethernet_fmhy_entry_t *e) {
    aethernet_fmhy_entry_t c;
    memset(&c, 0, sizeof(c));
    c.name = str_dup_fmhy(e->name);
    c.url = str_dup_fmhy(e->url);
    c.description = str_dup_fmhy(e->description);
    c.category = str_dup_fmhy(e->category);
    c.is_starred = e->is_starred;
    c.mirror_count = e->mirror_count;
    if (e->mirror_count > 0) {
        c.mirrors = (char **)calloc((size_t)e->mirror_count, sizeof(char *));
        if (c.mirrors) {
            for (int32_t i = 0; i < e->mirror_count; i++) c.mirrors[i] = str_dup_fmhy(e->mirrors[i]);
        } else {
            c.mirror_count = 0;
        }
    }
    return c;
}

// ─── growable list ───────────────────────────────────────

typedef struct {
    aethernet_fmhy_entry_t *items;
    int32_t count;
    int32_t cap;
} entry_list_t;

static bool list_push(entry_list_t *l, aethernet_fmhy_entry_t e) {
    if (l->count == l->cap) {
        int32_t ncap = l->cap ? l->cap * 2 : 16;
        aethernet_fmhy_entry_t *ni = (aethernet_fmhy_entry_t *)realloc(l->items, sizeof(*ni) * (size_t)ncap);
        if (!ni) return false;
        l->items = ni;
        l->cap = ncap;
    }
    l->items[l->count++] = e;
    return true;
}

// ─── parser ──────────────────────────────────────────────

// Find the first **[name](url)** in content; on success malloc name+url and set
// *bold_end to the offset after ")**". Returns true on success.
static bool find_bold(const char *content, char **out_name, char **out_url, size_t *bold_end) {
    size_t len = strlen(content);
    size_t from = 0;
    while (from < len) {
        const char *bp = strstr(content + from, "**[");
        if (!bp) break;
        size_t p = (size_t)(bp - content);
        size_t name_start = p + 3;
        const char *rb = strchr(content + name_start, ']');
        if (rb) {
            size_t name_end = (size_t)(rb - content);
            if (content[name_end + 1] == '(') {
                size_t url_start = name_end + 2;
                const char *rp = strchr(content + url_start, ')');
                if (rp) {
                    size_t url_end = (size_t)(rp - content);
                    if (content[url_end + 1] == '*' && content[url_end + 2] == '*') {
                        *out_name = trim_ndup(content + name_start, name_end - name_start);
                        *out_url = trim_ndup(content + url_start, url_end - url_start);
                        *bold_end = url_end + 3;
                        return true;
                    }
                }
            }
        }
        from = p + 3;
    }
    return false;
}

// Extract plain [name](url) URLs from region[0..rlen). Appends trimmed urls
// (malloc'd) to *list (grown via realloc); updates *n / *cap.
static void collect_plain_urls(const char *region, size_t rlen, char ***list, int32_t *n, int32_t *cap) {
    size_t from = 0;
    while (from < rlen) {
        const char *lb = memchr(region + from, '[', rlen - from);
        if (!lb) break;
        size_t p = (size_t)(lb - region);
        const char *rb = memchr(region + p + 1, ']', rlen - (p + 1));
        if (rb) {
            size_t name_end = (size_t)(rb - region);
            if (name_end + 1 < rlen && region[name_end + 1] == '(') {
                size_t url_start = name_end + 2;
                const char *rp = (url_start < rlen) ? memchr(region + url_start, ')', rlen - url_start) : NULL;
                if (rp) {
                    size_t url_end = (size_t)(rp - region);
                    char *u = trim_ndup(region + url_start, url_end - url_start);
                    if (u) {
                        if (*n == *cap) {
                            int32_t nc = *cap ? *cap * 2 : 4;
                            char **nl = (char **)realloc(*list, sizeof(char *) * (size_t)nc);
                            if (nl) { *list = nl; *cap = nc; }
                        }
                        if (*n < *cap) (*list)[(*n)++] = u; else free(u);
                    }
                    from = url_end + 1;
                    continue;
                }
            }
        }
        from = p + 1;
    }
}

// Strip [name](url) markdown links from text, keeping the name. Returns malloc'd.
static char *strip_links(const char *text) {
    size_t len = strlen(text);
    char *out = (char *)malloc(len + 1);
    if (!out) return NULL;
    size_t o = 0, from = 0;
    while (from < len) {
        const char *lb = memchr(text + from, '[', len - from);
        if (!lb) break;
        size_t p = (size_t)(lb - text);
        memcpy(out + o, text + from, p - from);
        o += p - from;
        const char *rb = memchr(text + p + 1, ']', len - (p + 1));
        bool consumed = false;
        if (rb) {
            size_t name_end = (size_t)(rb - text);
            if (name_end + 1 < len && text[name_end + 1] == '(') {
                size_t url_start = name_end + 2;
                const char *rp = (url_start < len) ? memchr(text + url_start, ')', len - url_start) : NULL;
                if (rp) {
                    size_t url_end = (size_t)(rp - text);
                    memcpy(out + o, text + p + 1, name_end - (p + 1)); // keep name
                    o += name_end - (p + 1);
                    from = url_end + 1;
                    consumed = true;
                }
            }
        }
        if (!consumed) {
            out[o++] = '[';
            from = p + 1;
        }
    }
    memcpy(out + o, text + from, len - from);
    o += len - from;
    out[o] = '\0';
    return out;
}

aethernet_fmhy_entry_t *aethernet_fmhy_parse_markdown(const char *markdown, int32_t *out_count) {
    entry_list_t list = {0};
    if (out_count) *out_count = 0;
    if (!markdown) return NULL;

    char *h1 = str_dup_fmhy("");
    char *h2 = str_dup_fmhy("");

    const char *cursor = markdown;
    while (*cursor || cursor == markdown) {
        const char *nl = strchr(cursor, '\n');
        size_t raw_len = nl ? (size_t)(nl - cursor) : strlen(cursor);
        // Trailing-whitespace-trimmed copy of the line.
        size_t line_len = raw_len;
        while (line_len > 0 && is_ws(cursor[line_len - 1])) line_len--;

        if (line_len > 0) {
            char *line = (char *)malloc(line_len + 1);
            if (line) {
                memcpy(line, cursor, line_len);
                line[line_len] = '\0';

                // Heading: 1-2 leading '#', then whitespace.
                size_t hashes = 0;
                while (line[hashes] == '#') hashes++;
                if ((hashes == 1 || hashes == 2) && (line[hashes] == ' ' || line[hashes] == '\t')) {
                    char *title = trim_ndup(line + hashes, line_len - hashes);
                    if (hashes == 1) { free(h1); h1 = title; free(h2); h2 = str_dup_fmhy(""); }
                    else { free(h2); h2 = title; }
                    free(line);
                    goto next_line;
                }

                // Bullet: optional leading ws, '*' or '-', required ws, content.
                size_t bi = 0;
                while (is_ws(line[bi])) bi++;
                if (line[bi] == '*' || line[bi] == '-') {
                    size_t after = bi + 1;
                    if (is_ws(line[after])) {
                        while (is_ws(line[after])) after++;
                        if (line[after] != '\0') {
                            const char *content = line + after;
                            bool is_starred = strstr(content, "\xE2\xAD\x90") != NULL; // ⭐
                            char *name = NULL, *url = NULL;
                            size_t bold_end = 0;
                            if (find_bold(content, &name, &url, &bold_end)
                                && url && url[0] != '\0' && url[0] != '#') {
                                size_t clen = strlen(content);
                                // Description after first " - " past the bold link.
                                char *description = NULL;
                                const char *sep_ptr = strstr(content + bold_end, " - ");
                                size_t desc_sep = sep_ptr ? (size_t)(sep_ptr - content) : (size_t)-1;
                                if (sep_ptr) {
                                    char *raw_desc = trim_ndup(content + desc_sep + 3, clen - (desc_sep + 3));
                                    if (raw_desc) {
                                        char *stripped = strip_links(raw_desc);
                                        free(raw_desc);
                                        if (stripped) {
                                            char *trimmed = trim_ndup(stripped, strlen(stripped));
                                            free(stripped);
                                            if (trimmed && trimmed[0] != '\0') description = trimmed;
                                            else free(trimmed);
                                        }
                                    }
                                }
                                // Mirrors between the bold link and the description.
                                size_t region_len = (sep_ptr ? desc_sep : clen) - bold_end;
                                char **mirrors = NULL;
                                int32_t mn = 0, mcap = 0;
                                collect_plain_urls(content + bold_end, region_len, &mirrors, &mn, &mcap);
                                // Drop mirrors equal to the primary url or starting with '#'.
                                int32_t kept = 0;
                                for (int32_t i = 0; i < mn; i++) {
                                    if (mirrors[i][0] != '\0' && strcmp(mirrors[i], url) != 0 && mirrors[i][0] != '#') {
                                        mirrors[kept++] = mirrors[i];
                                    } else {
                                        free(mirrors[i]);
                                    }
                                }
                                mn = kept;

                                // Category.
                                char *category;
                                if (h2[0] != '\0') {
                                    size_t cl = strlen(h1) + 3 + strlen(h2) + 1;
                                    category = (char *)malloc(cl);
                                    if (category) snprintf(category, cl, "%s / %s", h1, h2);
                                } else {
                                    category = str_dup_fmhy(h1);
                                }

                                aethernet_fmhy_entry_t e;
                                memset(&e, 0, sizeof(e));
                                e.name = name;
                                e.url = url;
                                e.description = description;
                                e.category = category;
                                e.is_starred = is_starred;
                                e.mirrors = mirrors;
                                e.mirror_count = mn;
                                if (!list_push(&list, e)) free_entry_fields(&e);
                                name = url = NULL; // ownership moved
                            }
                            free(name);
                            free(url);
                        }
                    }
                }
                free(line);
            }
        }

    next_line:
        if (!nl) break;
        cursor = nl + 1;
    }

    free(h1);
    free(h2);
    if (out_count) *out_count = list.count;
    return list.items;
}

// ─── tracker sources ─────────────────────────────────────

static const aethernet_fmhy_tracker_source_t kTrackerSources[] = {
    {"ngosang/trackerslist", "https://ngosang.github.io/trackerslist/trackers_all.txt", "Community-maintained list of all known public BitTorrent trackers."},
    {"XIU2/TrackersListCollection (all)", "https://raw.githubusercontent.com/XIU2/TrackersListCollection/master/all.txt", "Comprehensive tracker collection maintained by XIU2, updated daily."},
    {"XIU2/TrackersListCollection (best)", "https://raw.githubusercontent.com/XIU2/TrackersListCollection/master/best.txt", "Curated best-performing tracker subset from the XIU2 collection."},
    {"newtrackon (stable)", "https://newtrackon.com/api/stable", "Live-monitored stable tracker list from newtrackon.com."},
    {"openwebtorrent", "https://openwebtorrent.com/", "Free WebTorrent-compatible tracker for browser-based torrenting."},
};

const aethernet_fmhy_tracker_source_t *aethernet_fmhy_tracker_sources(int32_t *out_count) {
    if (out_count) *out_count = (int32_t)(sizeof(kTrackerSources) / sizeof(kTrackerSources[0]));
    return kTrackerSources;
}

// ─── service ─────────────────────────────────────────────

struct aethernet_fmhy_service {
    aethernet_fmhy_entry_t *entries;
    int32_t count;
};

aethernet_fmhy_service_t *aethernet_fmhy_service_new(void) {
    return (aethernet_fmhy_service_t *)calloc(1, sizeof(aethernet_fmhy_service_t));
}

void aethernet_fmhy_service_free(aethernet_fmhy_service_t *service) {
    if (!service) return;
    aethernet_fmhy_entries_free(service->entries, service->count);
    free(service);
}

void aethernet_fmhy_sync(aethernet_fmhy_service_t *service, const char *markdown) {
    if (!service) return;
    aethernet_fmhy_entries_free(service->entries, service->count);
    service->entries = aethernet_fmhy_parse_markdown(markdown, &service->count);
}

int32_t aethernet_fmhy_entry_count(aethernet_fmhy_service_t *service) {
    return service ? service->count : 0;
}

static aethernet_fmhy_entry_t *filter_copy(
    aethernet_fmhy_service_t *service, const char *category_filter, bool starred_only, int32_t *out_count) {
    if (out_count) *out_count = 0;
    if (!service) return NULL;
    bool has_filter = category_filter && category_filter[0] != '\0';

    entry_list_t out = {0};
    for (int32_t i = 0; i < service->count; i++) {
        const aethernet_fmhy_entry_t *e = &service->entries[i];
        if (starred_only && !e->is_starred) continue;
        if (has_filter && !contains_ci(e->category, category_filter)) continue;
        aethernet_fmhy_entry_t c = copy_entry(e);
        if (!list_push(&out, c)) free_entry_fields(&c);
    }
    if (out_count) *out_count = out.count;
    return out.items;
}

aethernet_fmhy_entry_t *aethernet_fmhy_browse(aethernet_fmhy_service_t *service, const char *category_filter, int32_t *out_count) {
    return filter_copy(service, category_filter, false, out_count);
}

aethernet_fmhy_entry_t *aethernet_fmhy_get_starred(aethernet_fmhy_service_t *service, const char *category_filter, int32_t *out_count) {
    return filter_copy(service, category_filter, true, out_count);
}

File: rust/src/fmhy.rs
// SPDX-License-Identifier: MIT
//! Free Media Heck Yeah (FMHY) content catalogue (Phase-2 extension).
//!
//! Propagated over the Aether mesh so offline peers benefit from entries fetched
//! by connected peers. Port of the C# reference (AetherNet.Fmhy): a markdown
//! parser for the FMHY single-page dump plus an in-memory catalogue. The parser
//! is hand-rolled (the crate has no `regex` dependency) but mirrors the reference
//! grammar exactly.

/// A single resource parsed from the FMHY directory.
#[derive(Clone, Debug)]
pub struct FmhyEntry {
    pub name: String,
    pub url: String,
    pub description: Option<String>,
    pub category: String, // "H1" or "H1 / H2"
    pub is_starred: bool,
    pub mirrors: Vec<String>,
}

impl FmhyEntry {
    /// All URLs: primary followed by any mirrors.
    pub fn all_urls(&self) -> Vec<String> {
        let mut v = Vec::with_capacity(1 + self.mirrors.len());
        v.push(self.url.clone());
        v.extend(self.mirrors.iter().cloned());
        v
    }
}

/// A known torrent tracker-list aggregator.
#[derive(Clone, Debug)]
pub struct TrackerSource {
    pub name: String,
    pub url: String,
    pub description: String,
}

/// The public FMHY single-page endpoint.
pub const FMHY_API_URL: &str = "https://api.fmhy.net/single-page";

/// Well-known public tracker-list aggregators bundled with this release.
pub fn built_in_tracker_sources() -> Vec<TrackerSource> {
    fn t(name: &str, url: &str, desc: &str) -> TrackerSource {
        TrackerSource { name: name.to_string(), url: url.to_string(), description: desc.to_string() }
    }
    vec![
        t("ngosang/trackerslist", "https://ngosang.github.io/trackerslist/trackers_all.txt", "Community-maintained list of all known public BitTorrent trackers."),
        t("XIU2/TrackersListCollection (all)", "https://raw.githubusercontent.com/XIU2/TrackersListCollection/master/all.txt", "Comprehensive tracker collection maintained by XIU2, updated daily."),
        t("XIU2/TrackersListCollection (best)", "https://raw.githubusercontent.com/XIU2/TrackersListCollection/master/best.txt", "Curated best-performing tracker subset from the XIU2 collection."),
        t("newtrackon (stable)", "https://newtrackon.com/api/stable", "Live-monitored stable tracker list from newtrackon.com."),
        t("openwebtorrent", "https://openwebtorrent.com/", "Free WebTorrent-compatible tracker for browser-based torrenting."),
    ]
}

// Heading: 1-2 leading '#', then whitespace, then title. Returns (level, title).
fn parse_heading(line: &str) -> Option<(usize, String)> {
    let b = line.as_bytes();
    let mut i = 0;
    while i < b.len() && b[i] == b'#' {
        i += 1;
    }
    if (i == 1 || i == 2) && i < b.len() && (b[i] == b' ' || b[i] == b'\t') {
        return Some((i, line[i..].trim().to_string()));
    }
    None
}

// Bullet: optional leading whitespace, '*' or '-', required whitespace, content.
fn parse_bullet(line: &str) -> Option<&str> {
    let trimmed = line.trim_start();
    let b = trimmed.as_bytes();
    if b.is_empty() || (b[0] != b'*' && b[0] != b'-') {
        return None;
    }
    let rest = &trimmed[1..];
    let content = rest.trim_start();
    if rest.len() == content.len() || content.is_empty() {
        return None; // require whitespace after the bullet, and non-empty content
    }
    Some(content)
}

// First **[name](url)** in `content`. Returns (name, url, byte offset after ")**").
fn find_bold_link(content: &str) -> Option<(String, String, usize)> {
    let mut from = 0;
    while let Some(rel) = content[from..].find("**[") {
        let p = from + rel;
        let name_start = p + 3;
        if let Some(nrel) = content[name_start..].find(']') {
            let name_end = name_start + nrel;
            if content[name_end..].starts_with("](") {
                let url_start = name_end + 2;
                if let Some(urel) = content[url_start..].find(')') {
                    let url_end = url_start + urel;
                    if content[url_end..].starts_with(")**") {
                        return Some((
                            content[name_start..name_end].trim().to_string(),
                            content[url_start..url_end].trim().to_string(),
                            url_end + 3,
                        ));
                    }
                }
            }
        }
        from = p + 3;
    }
    None
}

// Plain [name](url) links in `region`; returns their URLs.
fn plain_link_urls(region: &str) -> Vec<String> {
    let mut out = Vec::new();
    let mut from = 0;
    while let Some(rel) = region[from..].find('[') {
        let p = from + rel;
        if let Some(nrel) = region[p + 1..].find(']') {
            let name_end = p + 1 + nrel;
            if region[name_end..].starts_with("](") {
                let url_start = name_end + 2;
                if let Some(urel) = region[url_start..].find(')') {
                    let url_end = url_start + urel;
                    out.push(region[url_start..url_end].to_string());
                    from = url_end + 1;
                    continue;
                }
            }
        }
        from = p + 1;
    }
    out
}

// Replace [name](url) with name (strip residual markdown links from a description).
fn strip_links(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut from = 0;
    while let Some(rel) = text[from..].find('[') {
        let p = from + rel;
        out.push_str(&text[from..p]);
        if let Some(nrel) = text[p + 1..].find(']') {
            let name_end = p + 1 + nrel;
            if text[name_end..].starts_with("](") {
                let url_start = name_end + 2;
                if let Some(urel) = text[url_start..].find(')') {
                    let url_end = url_start + urel;
                    out.push_str(&text[p + 1..name_end]); // keep the name
                    from = url_end + 1;
                    continue;
                }
            }
        }
        out.push('[');
        from = p + 1;
    }
    out.push_str(&text[from..]);
    out
}

/// Parse a raw FMHY markdown string into a flat list of entries in document order.
pub fn parse_fmhy_markdown(markdown: &str) -> Vec<FmhyEntry> {
    let mut entries = Vec::new();
    let mut h1 = String::new();
    let mut h2 = String::new();

    for raw in markdown.split('\n') {
        let line = raw.trim_end_matches([' ', '\t', '\r']);
        if line.is_empty() {
            continue;
        }

        if let Some((level, title)) = parse_heading(line) {
            if level == 1 {
                h1 = title;
                h2 = String::new();
            } else {
                h2 = title;
            }
            continue;
        }

        let content = match parse_bullet(line) {
            Some(c) => c,
            None => continue,
        };
        let is_starred = content.contains('\u{2B50}'); // ⭐

        let (name, url, bold_end) = match find_bold_link(content) {
            Some(v) => v,
            None => continue,
        };
        if url.is_empty() || url.starts_with('#') {
            continue;
        }

        let mut description: Option<String> = None;
        let desc_sep = content[bold_end..].find(" - ").map(|r| r + bold_end);
        if let Some(sep) = desc_sep {
            let d = strip_links(content[sep + 3..].trim()).trim().to_string();
            if !d.is_empty() {
                description = Some(d);
            }
        }

        let mirror_region = match desc_sep {
            Some(sep) => &content[bold_end..sep],
            None => &content[bold_end..],
        };
        let mut mirrors = Vec::new();
        for mu in plain_link_urls(mirror_region) {
            let mu = mu.trim().to_string();
            if !mu.is_empty() && mu != url && !mu.starts_with('#') {
                mirrors.push(mu);
            }
        }

        let category = if !h2.is_empty() {
            format!("{h1} / {h2}")
        } else {
            h1.clone()
        };
        entries.push(FmhyEntry { name, url, description, category, is_starred, mirrors });
    }
    entries
}

/// In-memory FMHY catalogue, seeded optionally and updated via [`InMemoryFmhyCatalogueService::sync`].
#[derive(Default)]
pub struct InMemoryFmhyCatalogueService {
    entries: Vec<FmhyEntry>,
    last_synced_at_secs: Option<u64>,
}

impl InMemoryFmhyCatalogueService {
    pub fn new(seed: Vec<FmhyEntry>) -> Self {
        Self { entries: seed, last_synced_at_secs: None }
    }

    pub fn entry_count(&self) -> usize {
        self.entries.len()
    }

    pub fn last_synced_at_secs(&self) -> Option<u64> {
        self.last_synced_at_secs
    }

    /// Replace the catalogue from a fresh FMHY markdown string.
    pub fn sync(&mut self, markdown: &str) {
        self.entries = parse_fmhy_markdown(markdown);
        self.last_synced_at_secs = Some(
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_secs())
                .unwrap_or(0),
        );
    }

    /// All entries, optionally filtered by a case-insensitive category substring.
    pub fn browse(&self, category_filter: Option<&str>) -> Vec<FmhyEntry> {
        match category_filter {
            None | Some("") => self.entries.clone(),
            Some(cf) => {
                let cf = cf.to_lowercase();
                self.entries
                    .iter()
                    .filter(|e| e.category.to_lowercase().contains(&cf))
                    .cloned()
                    .collect()
            }
        }
    }

    /// Only starred entries, optionally category-filtered.
    pub fn get_starred(&self, category_filter: Option<&str>) -> Vec<FmhyEntry> {
        let cf = category_filter.filter(|s| !s.is_empty()).map(|s| s.to_lowercase());
        self.entries
            .iter()
            .filter(|e| {
                e.is_starred && cf.as_ref().map_or(true, |c| e.category.to_lowercase().contains(c))
            })
            .cloned()
            .collect()
    }

    /// The bundled tracker-list aggregators.
    pub fn get_tracker_sources(&self) -> Vec<TrackerSource> {
        built_in_tracker_sources()
    }
}
